security-framework/src/trust_settings.rs

```diff
@@ -0,0 +1,253 @@
+    Deny,
+
+    /// Neither trusted nor distrusted.
+    Unspecified
+}
+
+impl TrustSettingsForCertificate {
+    fn new(value: i64) -> TrustSettingsForCertificate {
+        match u32::try_from(value).ok() {
+            Some(kSecTrustSettingsResultTrustRoot) => TrustSettingsForCertificate::TrustRoot,
+            Some(kSecTrustSettingsResultTrustAsRoot) => TrustSettingsForCertificate::TrustAsRoot,
+            Some(kSecTrustSettingsResultDeny) => TrustSettingsForCertificate::Deny,
+            Some(kSecTrustSettingsResultUnspecified) => TrustSettingsForCertificate::Unspecified,
+            Some(_) | None => TrustSettingsForCertificate::Invalid,
+        }
+    }
+}
+
+/// Allows access to the certificates and their trust settings in a given domain.
+pub struct TrustSettings {
+    domain: Domain,
+}
+
+impl TrustSettings {
+    /// Create a new TrustSettings for the given domain.
+    ///
+    /// You can call `iter()` to discover the certificates with settings in this domain.
+    ///
+    /// Then you can call `tls_trust_settings_for_certificate()` with a given certificate
+    /// to learn what the aggregate trust setting for that certificate within this domain.
+    pub fn new(domain: Domain) -> TrustSettings {
+        TrustSettings { domain }
+    }
+
+    /// Create an iterator over the certificates with settings in this domain.
+    /// This produces an empty iterator if there are no such certificates.
+    pub fn iter(&self) -> Result<TrustSettingsIter> {
+        let array = unsafe {
+            let mut array_ptr: CFArrayRef = ptr::null_mut();
+
+            // SecTrustSettingsCopyCertificates returns errSecNoTrustSettings
+            // if no items have trust settings in the given domain.  We map
+            // that to an empty TrustSettings iterator.
+            match SecTrustSettingsCopyCertificates(self.domain.into(), &mut array_ptr) {
+                errSecNoTrustSettings => {
+                    Ok(CFArray::from_CFTypes(&[]))
+                },
+                errSecSuccess => {
+                    Ok(CFArray::<SecCertificate>::wrap_under_create_rule(array_ptr))
+                },
+                err => Err(Error::from_code(err)),
+            }?
+        };
+
+        Ok(TrustSettingsIter {
+            index: 0,
+            array,
+        })
+    }
+
+    /// Returns the aggregate trust setting for the given certificate.
+    ///
+    /// This tells you whether the certificate should be trusted as a TLS
+    /// root certificate.
+    pub fn tls_trust_settings_for_certificate(&self, cert: &SecCertificate)
+        -> Result<TrustSettingsForCertificate> {
+        let trust_settings = unsafe {
+            let mut array_ptr: CFArrayRef = ptr::null_mut();
+            let cert_ptr = cert.as_CFTypeRef() as *mut _;
+            match SecTrustSettingsCopyTrustSettings(cert_ptr, self.domain.into(), &mut array_ptr) {
+                errSecNoTrustSettings => {
+                    Ok(CFArray::from_CFTypes(&[]))
```

Comment:
The docs say " An empty Trust Settings array is definitely not the same as *no* Trust Settings, which means "this cert must be verified to a known trusted cert". , but an empty array here will fall back to trust!

Reply:
Yeah, it seems reasonable to return None here if there were no errors but, equally, no specific settings. I'll do that.